src/python/pants/option/config.py

```diff
@@ -230,10 +244,187 @@ def get_value(self, section: str, option: str) -> Optional[str]:
   def options(self, section: str) -> List[str]:
     return self.parser.options(section)
 
+  @property
   def defaults(self) -> Mapping[str, str]:
     return self.parser.defaults()
 
 
+_TomlPrimitve = Union[bool, int, float, str]
+_TomlValue = Union[_TomlPrimitve, List[_TomlPrimitve]]
+
+
+@dataclass(frozen=True)
+class _TomlValues(_ConfigValues):
+  values: Dict[str, Any]
+
+  @staticmethod
+  def _is_an_option(option_value: Union[_TomlValue, Dict]) -> bool:
+    """Determine if the value is actually an option belonging to that section.
+
+    A value that looks like an option might actually be a subscope, e.g. the option value
+    `java` belonging to the section `cache` could actually be the section `cache.java`, rather
+    than the option `--cache-java`.
+
+    We must also handle the special syntax of `my_list_option.add` and `my_list_option.remove`.
+    """
+    return (
+      not isinstance(option_value, dict) or "add" in option_value or "remove" in option_value
+    )
+
+  @staticmethod
+  def _section_explicitly_defined(section_values: Dict) -> bool:
+    """Determine if the section is truly a defined section, meaning that the user explicitly wrote
+    the section in their config file.
+
+    For example, the user may have explicitly defined `cache.java` but never defined `cache`. Due
+    to TOML's representation of the config as a nested dictionary, naively, it would appear that
+    `cache` was defined even though the user never explicitly added it to their config.
+    """
+    at_least_one_option_defined = any(
+      _TomlValues._is_an_option(section_value) for section_value in section_values.values()
+    )
+    blank_section = len(section_values.values()) == 0
```

Comment:
Absolutely a nit: It took me a bit to understand why a blank section must be explicitly defined. May I suggest a comment: 
```suggestion
    # A section that has no options cannot have been implicitly created by an option.
    blank_section = len(section_values.values()) == 0
```